[Buttons block]: Add gap and vertical margin support

stacimc opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->
Fixes: #29098 

Related/Depends on: #28365
* In particular, pending the discussion [here](https://github.com/WordPress/gutenberg/issues/28356#issuecomment-915072804) about how to display the gap controls in the UI

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->
Makes use of the new Gap block support (#33991) in the Buttons block to allow control over the space between buttons. At the moment horizontal and vertical gaps are not handled separately. It also adds vertical margin support to allow adjusting the margin above and below the Buttons container.

As with Columns in #34456, this is intended as a stop-gap to get the Gap block support used quickly in Buttons, while we separately look into making the block use the `flex` layout. My reasoning is that this approach works with vertically oriented buttons, while flex layouts currently only support horizontal layouts.

It also fixes a few pre-existing buttons bugs:
* Vertically oriented buttons should not factor the gap into their width calculation
* Vertically oriented center-aligned buttons that set 100% width were not filling the row; this is now fixed

### Major Tradeoffs
As we move from using top & bottom margins to gap for controlling vertical spacing, there will be some styling regressions in themes that are currently tweaking those margin values.
* **Vertical spacing between the Buttons container and adjacent blocks may be affected.** This is because individual buttons no longer have top/bottom margin, which previously would have 'padded' the Buttons container.
* **Row gap may be affected**. Themes may have previously increased the row-gap by adding extra margin to buttons. Now this margin is overwritten to `0`, themes will need to tweak the `row-gap` value instead.

In particular here are two known bugs:
* Twenty Twenty One forces `margin: 0` on the Buttons container and adds extra margin to the `:first-child` button. With this PR applied, this causes _all space_ between the Buttons container and adjacent blocks to disappear (unless the other block has its own margin), and vertically offsets the first button. https://core.trac.wordpress.org/ticket/54250
* Twenty Twenty uses margins to increase the row-spacing of buttons on the frontend. With this PR applied, the margins are overwritten and this vertical spacing shrinks. https://core.trac.wordpress.org/ticket/54251

### Notes

* Button widths take the gap into consideration. This means if you have two buttons in a horizontal orientation, both set to 50% width, they will fit on the same line and each take up 50% of the row _minus the gap in between them_. This can look odd if you set a huge gap because a button set to 50% width could end up smaller than expected, but is intuitive with more "normal" gap selections.
<img width="520" alt="Screen Shot 2021-09-08 at 11 20 26 AM" src="https://user-images.githubusercontent.com/63313398/132564111-5b3f21e6-ec1f-4a05-9a0d-aa4b2a266982.png">

* When the "Space between items" alignment option is selected, buttons will be spaced equally across their row as expected -- so effectively, the column gap is _at least_ as big as the selected gap, but may be bigger if needed to make the buttons fit to the row. The row gap is exactly what is specified in the user selection. Ie in this screenshot I set a 3px gap: The horizontal gap between buttons is larger than 3px in order to evenly space them on the row; the vertical gap is exactly 3px.
<img width="615" alt="Screen Shot 2021-09-08 at 11 23 14 AM" src="https://user-images.githubusercontent.com/63313398/132564413-8ccf51b5-ca9b-4669-859b-be35622a1c5d.png">

* If you choose the vertically oriented variation of Buttons, Gap is still supported but of course only affects the vertical spacing between buttons
<img width="619" alt="Screen Shot 2021-09-08 at 2 04 55 PM" src="https://user-images.githubusercontent.com/63313398/132585421-a2fcb7da-6ac1-43f8-b404-60d58e8cc221.png">




## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

Manually. Some areas to focus:
* Changing button widths
* Changing items justification
* Vertical as well as horizontal orientation
* Full/Wide/Normal widths
* Buttons inside a container with a flex layout, like the Group block
* Verifying all styling applies on the frontend
* Set a gap value through `theme.json`
* Set a gap value through global styles

## Screenshots <!-- if applicable -->

![button-gap](https://user-images.githubusercontent.com/63313398/132585074-91602c8d-90bb-419a-8bfb-851c23cfe296.gif)

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

stacimc committed:
Add gap support to Buttons

stacimc committed:
Use css var and remove additional top and bottom margin

ZebulanStanphill commented on packages/block-library/src/buttons/style.scss:
```diff
@@ -5,7 +5,8 @@ $blocks-block__margin: 0.5em;
 	flex-direction: row;
 	flex-wrap: wrap;
-	column-gap: $blocks-block__margin;
+	column-gap: var(--wp--style--block-gap, $blocks-block__margin);
+	row-gap: var(--wp--style--block-gap, $blocks-block__margin);
```
This means the same thing:
```suggestion
	gap: var(--wp--style--block-gap, $blocks-block__margin);
```

stacimc replied:
Yep, sorry this is in a muddled state in progress as I'm exploring a couple different options. This in particular is a vestige from when I was working with independent row and column gaps and definitely will be changed.

stacimc committed:
Remove unneeded CSS and account for gap in width calculations

andrewserong commented on packages/block-library/src/button/style.scss:
```diff
@@ -50,57 +50,44 @@ $blocks-block__margin: 0.5em;
-				flex-basis: 100%;
-			}
+		&.wp-block-button__width-100 {
+			width: auto;
```
I know this isn't a change in your PR, but I noticed while testing with TT1-blocks that when using the vertical variation, setting to 100% via `width: auto` results in the block being only as wide as its contents. Should this value be `width: 100%` instead? It appears to fix it for me in TT1-blocks.

![buttons-vertical-100-sml](https://user-images.githubusercontent.com/14988353/132634090-5b815f01-2834-4f07-8862-12a842268abe.gif)

stacimc replied:
Oh, really good catch! Seems like this is only happening for vertically oriented _and_ center-aligned buttons, but that doesn't look good at all! Adding a fix in, thank you.

andrewserong commented on packages/block-library/src/button/editor.scss:
```diff
@@ -1,9 +1,3 @@
-.wp-block[data-align="center"] > .wp-block-button {
```
I think this rule might still be required for the legacy standalone Button block if it appears outside of a Buttons container.

To get the legacy button block in there, If you add a paragraph to your post, then switch to the code editor view, and insert the following (a single Button block) at the end of the post content, then it'll expose the alignment controls for the standalone button:

```
<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link">Button A</a></div>
<!-- /wp:button -->
```

The left and right alignment appears to be working the same as on trunk, however the center alignment in the editor looks like this if we remove this CSS rule:

![image](https://user-images.githubusercontent.com/14988353/132635995-c216e05b-30ea-4b29-a48f-ea1a047901c0.png)

stacimc replied:
Yep, missed this one. Thanks!

andrewserong reviewed (commented):
Great work here @stacimc and thanks for the detailed PR description, there sure are a lot of variables when it comes to styling and positioning Button(s) blocks!

I've left a comment about the legacy standalone Button block (when it's rendered without a Buttons container), and a comment about the 100% width for the vertical variation.

Another slight oddity for the vertical variation, which isn't really caused by your PR because it exists on trunk, is that we're subtracting the block gap from the width of the 25%, 50% and 75% buttons. I think this works great in the horizontal layout as you've described (we expect the editor to figure out the appropriate balance with the gap when we set widths), however with the vertical layout when adjusting the gap, it seems unexpected that the gap would affect the widths... would it be worth adding rules like the following, or do you think it'd get too messy?

```
.wp-block-buttons.is-vertical > .wp-block-button {
  &.wp-block-button__width-25 {
    width: 25%;
  }
  &.wp-block-button__width-50 {
    width: 50%;
  }
  &.wp-block-button__width-75 {
    width: 75%;
  }
}
```

Other than that, I think this is looking really good, and even without refactoring the CSS all that much, by adding in the gap variable, you've been able to get rid of quite a bit of CSS which neatens it a fair bit. I'm particularly excited that we'll be able to tweak layouts like the following with gap support, though, so I think it's going to be a great feature:

![buttons-gap-sml](https://user-images.githubusercontent.com/14988353/132636977-dfc1d4f0-b684-40e6-adef-f6831f9bb814.gif)

stacimc committed:
Support center alignment for legacy buttons

stacimc committed:
Fix 100% width for center-aligned vertical buttons

stacimc committed:
Do not factor in gap when calculating width for vertical buttons

stacimc committed:
Clean up margin-based width calculations now that we're using gap

stacimc commented:
> Another slight oddity for the vertical variation, which isn't really caused by your PR because it exists on trunk, is that we're subtracting the block gap from the width of the 25%, 50% and 75% buttons.

Good point -- agreed, I would not expect gap to be accounted for in the widths of vertically oriented buttons. I'm not worried about adding in a few extra rules considering how much we're able to clean up with the rest of this PR :) Happy to add that in while I'm here!

andrewserong commented on packages/block-library/src/button/style.scss:
```diff
@@ -50,58 +49,35 @@ $blocks-block__margin: 0.5em;
 
 	&.wp-block-button__width-25 {
-		width: calc(25% - #{ $blocks-block__margin });
+		width: calc(25% - (var(--wp--style--block-gap, $blocks-block__margin) * 0.75));
```
It looks like to render the Sass variable `$blocks-block__margin` in this context (nested within var within a calc) we need to wrap it in the [Sass interpolation syntax](https://sass-lang.com/documentation/breaking-changes/css-vars) where the variable is wrapped in `#{}`, so for example this line would become:

```suggestion
		width: calc(25% - (var(--wp--style--block-gap, #{$blocks-block__margin}) * 0.75));
```

stacimc replied:
Thanks as always for the thorough testing -- I should've tested on a non-blocks theme! Updated 💯

andrewserong reviewed (commented):
Thanks for fixing up those issues @stacimc! That's testing nicely in TT1-Blocks now, and I tested in the site editor, too 👍 

I did a bit more testing today and tried out themes without a `theme.json` to test the fallback values, and noticed that it looks like Sass needs to use the [interpolation syntax](https://sass-lang.com/documentation/interpolation) `#{}` to inject the Sass variable when used within a `var()`.

To test, I switched my site to use the Maywood theme, then duplicated my test Buttons block and then removed the `gap` value so that the block doesn't have access to any `--wp--style--block-gap` and will use the `var`'s fallback value.

Here's a before / after with the interpolation added, so I think that's all that's needed to get this working on non-theme.json themes 🤞

| WIthout interpolation syntax | With interpolation syntax |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/132800198-b038f5c0-636a-4b1b-b84d-ec49f0371aed.png) | ![image](https://user-images.githubusercontent.com/14988353/132800238-34f603fe-4e4e-4e72-92ed-cfde5f66cfa9.png) |

Other than that, these changes are all testing well for me! Assuming others are fine with the change, I think the final step will be to see [where the discussion](https://github.com/WordPress/gutenberg/issues/28356#issuecomment-915072804) arrives at in regards to switching on the block gap and where we expose the control in the UI. Once we arrive at a conclusion, hopefully we'll then be able to get this and the [gallery](https://github.com/WordPress/gutenberg/pull/34608) and [columns](https://github.com/WordPress/gutenberg/pull/34630) blocks opt-in PRs merged.

jasmussen commented:
Thanks for this, gap feels like such a win here. 

These two:

> Another slight oddity for the vertical variation, which isn't really caused by your PR because it exists on trunk, is that we're subtracting the block gap from the width of the 25%, 50% and 75% buttons. I think this works great in the horizontal layout as you've described (we expect the editor to figure out the appropriate balance with the gap when we set widths), however with the vertical layout when adjusting the gap, it seems unexpected that the gap would affect the widths... would it be worth adding rules like the following, or do you think it'd get too messy?

> Button widths take the gap into consideration. This means if you have two buttons in a horizontal orientation, both set to 50% width, they will fit on the same line and each take up 50% of the row minus the gap in between them. This can look odd if you set a huge gap because a button set to 50% width could end up smaller than expected, but is intuitive with more "normal" gap selections.

Do we still need the `calc` rules when applying those widths? I had been my hope that they were necessary mostly to account for the margins. I think I had this discussion with @kjellr as well, where we went back and forward on how and when to start supporting gap here.

In either case, setting two buttons both to 50% width, I'd expect that to be 50% of the _allotted space_, i.e. if there's a 50% wide gap, then actually the buttons are 33% of the full container space. 

On a separate note, rows with multiple buttons is likely an edgecase, but it would be nice to upgrade the gap control to support axial configuration for the cases where you want a narrow horizontal gap and a tall vertical gap.

jasmussen commented on packages/block-library/src/button/editor.scss:
```diff
@@ -63,7 +63,6 @@
 	.is-selected & {
 		height: auto;
 		overflow: visible;
-		margin-top: $grid-unit-20;
```
Do you know what this margin was for?

stacimc replied:
It was added [here](https://github.com/WordPress/gutenberg/pull/10128/files), looks like it was applied to the `LinkControl` in the block toolbar. It's been refactored since then and I'm fairly confident this isn't getting used at all anymore, actually 🤔

I have tested adding/removing a button's link and the controls look good to me.

jasmussen commented on packages/block-library/src/button/style.scss:
```diff
@@ -34,7 +34,6 @@ $blocks-block__margin: 0.5em;
 	}
 }
 
-// Increased specificity needed to override margins
```
I guess the margins are gone, but is the specificity still needed for the gap?

stacimc replied:
I think we can remove this, yep. Thanks!

jasmussen commented on packages/block-library/src/button/style.scss:
```diff
@@ -50,58 +49,35 @@ $blocks-block__margin: 0.5em;
 }
 
-// If the browser supports column-gap, use that instead of the margins above.
-@supports ( column-gap: #{ $blocks-block__margin } ) {
```
Speaking of browser support, it's so nice to see the important ones green in [caniuse](https://caniuse.com/?search=gap). But I also recall something about Safari that was context for adding this check in the first place. Do you recall, @kjellr?

stacimc commented:
> Do we still need the `calc` rules when applying those widths? I had been my hope that they were necessary mostly to account for the margins. I think I had this discussion with @kjellr as well, where we went back and forward on how and when to start supporting gap here.

I was not able to get this to work without the `calc` rules, neither with `width` nor `flex-basis`. I played around with this today and could come up with nothing simpler than the `calc`, although I'm by no means a CSS expert. If it's possible I'd love to know how, it would be great to clean up even more rules here!

> In either case, setting two buttons both to 50% width, I'd expect that to be 50% of the _allotted space_, i.e. if there's a 50% wide gap, then actually the buttons are 33% of the full container space.

@jasmussen I think we're on the same page here and I'm just confused by the % gap, so making sure I understand: if I have say a 400px-wide row containing two buttons, each set to 50% width, and an explicitly 200px gap, then I would expect the buttons to each be 100px or _25%_ of the full container space.

I would still expect this to be true if the gap was set as a percentage -- so if I set the gap to 50%, the gap itself is 50% of the _container_ (200px), and then set my 2 buttons to 50% they would each take up 50% of the _remaining_ allotted 200px, rather than being treated like a ratio to the gap. Is that what you're thinking?

<img width="620" alt="Screen Shot 2021-09-10 at 4 29 15 PM" src="https://user-images.githubusercontent.com/63313398/132928045-6f74c0c3-7520-4dfe-a416-3c469ea8413c.png">

> On a separate note, rows with multiple buttons is likely an edgecase, but it would be nice to upgrade the gap control to support axial configuration for the cases where you want a narrow horizontal gap and a tall vertical gap.

Definitely ➕ ! I think the block support initially [explored that](https://github.com/WordPress/gutenberg/pull/33991) but it was decided that we should move that to a follow up. I like the idea of being able to control these separately for buttons.

>  I think the final step will be to see where the discussion arrives at in regards to switching on the block gap and where we expose the control in the UI. Once we arrive at a conclusion, hopefully we'll then be able to get this and the gallery and columns blocks opt-in PRs merged.

Thanks, I've updated the PR description to reflect this 👍

stacimc committed:
Interpolate vars and remove unneeded selector

paaljoachim commented:
I quickly browsed through the above comments.
I am wondering if gap support should contain a feature to unlock top, right, bottom and left so one can choose to have gaps in either direction. Similar to turning unlocking border radius to defining which side to use a radius.

stacimc committed:
Add top/bottom margin support to Buttons

stacimc committed:
Override block editor margins in non-blocks themes

stacimc commented:
> I am wondering if gap support should contain a feature to unlock top, right, bottom and left 

I agree, this would be really useful. #34529 is tracking the addition of axial gap support that would allow separate row/column gap configuration.

On a related note I think I'm going to have to do some more tinkering on this for backwards compatibility in non-blocks themes. The substantially smaller column spacing as compared to row spacing is one issue:

<img width="581" alt="Screen Shot 2021-09-16 at 2 17 06 PM" src="https://user-images.githubusercontent.com/63313398/133701411-3bc6eae6-2a6e-4967-9999-cb7a44f30dba.png">

This PR as it currently stands will also shrink the space between the Buttons container and blocks above/below it. I'm afraid we might not be able to get away with simplifying all the margin code after all.

stacimc committed:
Implement row gap with margins for backwards compatibility

stacimc commented:
This has been updated to implement the row gap using `margin`. This is unfortunately necessary to avoid backwards compatibility issues with classic themes such as Twenty Twenty, which style Buttons spacing by overriding the top and bottom `margin` on individual buttons.

The styling rules use `--wp--style--block-gap` and fallback to the existing default margin; this can cause styling regressions when there is no user-selected blockGap and it picks up the global variable instead. #35209 will fix this for classic themes, but as far as I can tell this is an unavoidable problem on blocks themes that don't explicitly opt out of both the gap block support _and_ the global spacing config (by setting `blockGap: null` in their `theme.json` settings).

What that means specifically is that on TT1 blocks, for example, on trunk we have a `0.5em` column-gap and `10px` margins (so `20px` row gap). With this PR applied, both column and row gap will adjust to the `24px` global gap by default:
| Before | After|
| --- | --- |
| <img width="652" alt="Screen Shot 2021-09-30 at 11 14 20 AM" src="https://user-images.githubusercontent.com/63313398/135530018-90ed1d4b-71d1-4029-8155-a8e708beb38a.png"> | <img width="653" alt="Screen Shot 2021-09-30 at 11 16 37 AM" src="https://user-images.githubusercontent.com/63313398/135530039-eed3235c-f00c-4ff1-9b7e-c8c382f241a6.png"> | 

Another problem caused by using `margin` to create the row gap is that the new gap spacing config already applies a `margin-top: var( --wp--style--block-gap)` to the Buttons container, with high specificity. Since the buttons have their own top margin, we effectively end up with 1.5 the size of the gap above the container:



| <img width="659" alt="Screen Shot 2021-09-30 at 3 00 23 PM" src="https://user-images.githubusercontent.com/63313398/135535875-81fd998d-4d94-48e7-99f6-198ba6bc8500.png"> |
|:--:|
| For dramatic effect, here `blockGap` is `120px`. So there is a `120px` margin-top applied to the Buttons container, and a `60px` margin-top and margin-bottom on each individual button => `180px` gap between the actual buttons and preceding content. |

Using the margin block support, the user can control the margin applied to the container itself but not to the individual buttons. This would be easily remedied if we could just use `row-gap`, but then we're back to backwards compatibility issues. 🤔

stacimc commented:
This PR is at a bit of an impasse due to some tradeoffs with respect to backwards compatibility. @jasmussen I'm curious how you feel about some of the priorities here?

One big tradeoff is the use of `margin` to implement row spacing instead of `row-gap`, which I'm doing to preserve backwards compatibility in classic themes (more details in my [above comment](https://github.com/WordPress/gutenberg/pull/34546#issuecomment-931704914))-- but which causes increased spacing between the Buttons container and adjacent blocks on blocks themes. It also means that even with added margin block support, a user can't fully control the spacing between individual buttons and preceding blocks, because there's a margin on the button blocks themselves. It's not clear to me that's worth the trade.

There's also the issue of axial gap support. Since that hasn't been implemented just yet (#34529), we get a bit of a styling regression on themes that opt into the block support, as there's no way to preserve existing styles where column gap is smaller than row gap.

Finally the unified variable approach for `blockGap` will inherently cause some regression for themes that opt in, because the global `--wp--style--block-gap` will get picked up.

jasmussen commented:
Backwards compatibility with regards to sufficient browser support is important, and I do recall there being an issue with `gap` in a particular version of Safari, however that's been addressed and support is looking good to me: https://caniuse.com/#search=gap

> This has been updated to implement the row gap using margin. This is unfortunately necessary to avoid backwards compatibility issues with classic themes such as Twenty Twenty, which style Buttons spacing by overriding the top and bottom margin on individual buttons.

I don't think backwards compatibility with themes is something that should block forward movement on features that otherwise vastly improve the user experience. It is important to note changes like these in changelogs and dev notes, and it's important to ticket the changes. And it's a balance of course, if the change affects every existing theme, it's one thing. If it affects a single or two themes, it's quite another.

For example, TwentyTwentyOne did the very best it could with the Navigation block in the state it was in when the theme launched. It was frankly impressive. However it had to add a great deal of CSS to make things work, and that CSS is now technical debt. We had to move forward with the navigation block, despite the issues that were caused. This ticket details those issues, and there's now a patch to fix TT1: https://core.trac.wordpress.org/ticket/53220

Can we get a sense of how big the problem is? I do recognize the problem of providing blanket top/bottom block margins, often by targetting `[data-block]`. Those rules almost all predate horizontal blocks like Buttons, Social Icons or even Navigation, and mostly those rules are editor style problems only. 

If the problem is widespread enough, it might be worth looking at creating some fallback rules and put them in https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-post/src/classic.scss, so that the appearance is okay. What do you think?

stacimc commented:
> I do recognize the problem of providing blanket top/bottom block margins, often by targetting [data-block]. Those rules almost all predate horizontal blocks like Buttons, Social Icons or even Navigation, and mostly those rules are editor style problems only.

As far as classic themes go, this appears to be the most common issue (Seedlet, Varia, Maywood, Twenty Twenty One, etc), but (as I'm glad you pointed out, @jasmussen) only affects the editor. These margins can be overridden in the button CSS and I suppose that would actually be a win, since we'd be bringing the editor in line with the frontend styles 🎉 

Less commonly, there are some theme styles that have highly specific selectors to add margins. Twenty Twenty One has an `.editor-styles-wrapper [data-block].wp-block-buttons .wp-block-button:first-child` selector which adds extra margin to the first button, and Twenty Twenty does something similar. Twenty Twenty also uses margin to increase the row spacing on the frontend.

| <img width="643" alt="Screen Shot 2021-10-04 at 12 09 53 PM" src="https://user-images.githubusercontent.com/63313398/135941783-fe211951-fb04-442b-b100-d111ab538ef2.png"> |
|:--:|
| _Twenty Twenty One: top margin applied to the first button in the editor_ |

| <img width="633" alt="Screen Shot 2021-10-04 at 11 23 52 AM" src="https://user-images.githubusercontent.com/63313398/135941827-4c2b271f-e01d-4ad1-b33e-28aaa21aef7e.png"> |
|:--:|
| _Twenty Twenty: margin is added on the frontend in addition to the row-gap_ |

The TT1 issue is visually very buggy (although it only affects the editor). I'm tempted to throw a `margin: 0 !important` on the button div; the tradeoff would be reducing that row gap in Twenty Twenty, and any other theme that does something similar. I'm not sure how to get a good idea of how many themes do something like that. Maybe it would be sufficient to note that themes can use `row-gap` within their stylesheets moving forward?

For blocks themes, the main issue is just that the gap support does not currently have an option for axial support, so for TT1-blocks for example the column gap has to grow to match the size of the row gap. My concern with increasing the column gap is that there may be sites with multiple buttons that used to fit in a single row, which now overflow into multiple rows. This would apply to any blocks theme that so much as enables the global gap spacing config, even if the block support itself is turned off.  I've started looking at adding axial support (#35301) separately.

So TLDR I think there are two big questions:

1. Can we use `row-gap` instead of margin, perhaps with the addition of `margin: 0 !important`, and rely on existing themes to update their styles accordingly?
2. Do we need to add axial support in the first iteration, or can it be a follow up?

andrewserong commented on packages/block-library/src/buttons/style.scss:
```diff
@@ -24,13 +22,8 @@ $blocks-block__margin: 0.5em;
-			margin-right: 0;
-		}
+		margin-right: 0;
+		margin-bottom: calc(var(--wp--style--block-gap, #{ $blocks-block__margin}) * 0.5);
```
Without a corresponding `margin-top` it looks like this results in a half-height gap on the front end. It currently looks fine in the editor due to the `margin-top` provided in `editor.scss`.

stacimc replied:
This and the related comment should be fixed by moving to gap for rows.

There is a bit of change here as mentioned earlier in the thread, since we're moving to row-gap we potentially shrink the gap between the entire Buttons block and adjacent blocks, which we previously got from the extra top/bottom margin on the buttons within the container.

For blocks themes, the addition of the margin block support will make this spacing easily user-configurable. For classic themes I imagine themes will need to override `margin` on the Buttons container itself if they want to increase this spacing.

andrewserong commented on packages/block-library/src/buttons/editor.scss:
```diff
@@ -1,18 +1,17 @@
+	// Specificity needed to override editor auto block margins.
 	> .wp-block {
-		// Override editor auto block margins.
-		margin-left: 0;
```
It looks like we might still need this rule for classic themes, which sets `.editor-styles-wrapper .wp-block` to `0` for margin left and right [here](https://github.com/wordpress/gutenberg/blob/7f97d5c3da65c73ea026171e3d9335c087ab2dde/packages/edit-post/src/classic.scss#L5-L5)

Left justified Buttons block in TwentyTwentyOne:

| Removing this rule | With this rule restored / set to 0 |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/135969973-5cf5d9c3-8a80-44b5-9d5a-6a623f3fb625.png) | ![image](https://user-images.githubusercontent.com/14988353/135970003-17f0188c-b95f-4905-9aa8-d93c79289c7d.png) |

stacimc replied:
Thanks, this should be fixed as we go back to using `gap` and setting all the margins to 0

andrewserong replied:
Unfortunately, I think we might still need this pesky left/right margin rule for classic themes. With TwentyNineteen applied, and a Buttons block set to "Justify items right", while the block is selected, the appender is rendered next to the buttons and receives the `wp-block` class. The margin-left/right from `classic.css` then creates too much room surrounding the appender, which visually makes the buttons appear as though they're left aligned:

![image](https://user-images.githubusercontent.com/14988353/136150752-bcdd293a-23a2-48c9-9626-844fa1a70752.png)

When de-selecting the Buttons block, it then appears to be aligned correctly:

![Kapture 2021-10-06 at 17 22 20](https://user-images.githubusercontent.com/14988353/136151019-b873c047-59fa-46ca-9da9-57bb6124a5b4.gif)

andrewserong replied:
I can reproduce this in TwentyTwenty and TwentyNineteen, here's a video from TwentyTwenty:

https://user-images.githubusercontent.com/14988353/136295389-b93a3873-c3bf-49a8-99f5-49b7f49d599f.mp4

andrewserong replied:
Thanks for fixing that up, this is testing nicely in the editor on classic themes now!

andrewserong commented on packages/block-library/src/buttons/style.scss:
```diff
@@ -116,7 +75,7 @@ $blocks-block__margin: 0.5em;
 		margin-left: auto;
 		margin-right: auto;
-		margin-bottom: $blocks-block__margin;
+		margin-bottom: calc(var(--wp--style--block-gap, #{ $blocks-block__margin}) * 0.5);
```
Same note as above, it looks like there isn't a corresponding `margin-top` so this margin appears to be half-height?

andrewserong reviewed (commented):
Thanks for laying out the current issues clearly @stacimc! I did a little bit of extra testing with classic themes and added a couple more comments.

> My concern with increasing the column gap is that there may be sites with multiple buttons that used to fit in a single row, which now overflow into multiple rows.

I suppose one thing that mitigates this slightly is if users set explicit widths for the buttons, then the change in gap won't affect whether they wrap to a new line?

For the two big questions:

> Can we use row-gap instead of margin, perhaps with the addition of margin: 0 !important, and rely on existing themes to update their styles accordingly?

The more that I look at the Button styles this idea seems really appealing, to be able avoid some of the more complex (and potentially brittle) CSS rules we have.

> Do we need to add axial support in the first iteration, or can it be a follow up?

Because axial support will likely be quite complex and fiddly to get right, I'm a +1 for not going with axial support in the first round of us opting blocks in to the gap support, so that we can then look at it in follow-ups. While it'd be great to be able to have a different row gap versus column gap, my gut feeling is that in most cases, buttons usually appear on sites in a single horizontal row of buttons, or a single column of stacked buttons. I could be wrong, though!

For those cases where folks need a short-term work around, I suppose they could add two Buttons blocks and use margins to tweak the vertical gap?

jasmussen commented:
Thanks for all the context!

> These margins can be overridden in the button CSS and I suppose that would actually be a win, since we'd be bringing the editor in line with the frontend styles 🎉

> Less commonly, there are some theme styles that have highly specific selectors to add margins. Twenty Twenty One has an .editor-styles-wrapper [data-block].wp-block-buttons .wp-block-button:first-child selector which adds extra margin to the first button, and Twenty Twenty does something similar. Twenty Twenty also uses margin to increase the row spacing on the frontend.

The very careful balance we have to tread here, though, is balancing the need to support existing WordPress themes, with avoiding technical debt or sub-optimal solutions in the block editor, if we can. In the case of TwentyTwentyOne, as noted it did the best it could with blocks as they existed on release, but it is now very buggy. I feel somewhat strongly that _we fix this in the theme, not in the block editor_. [A patch](https://core.trac.wordpress.org/ticket/53220) already fixes some of the margin issues around the navigation block. For any bugs with the Buttons block, a trac ticket and/or patch to the theme is the best path forward.

For example, the first-child rule in TwentyTwentyOne is problematic also for Social Icons and Navigation, and if we fix it in the block editor, it means we break the ability to style blocks via global styles, which requires the block CSS to have low specificity. For that reason, I would caution against fixing theme bugs by increasing specificity in blocks. 

A good approach would be to start with optimizing for a theme such as "Empty Theme" from https://github.com/WordPress/theme-experiments. That way you make sure you're not fixing theme bugs.

apeatling commented:
> I feel somewhat strongly that we fix this in the theme, not in the block editor.

+1 to fixing this margin issue in the theme(s) as @jasmussen mentioned above.

> Because axial support will likely be quite complex and fiddly to get right, I'm a +1 for not going with axial support in the first round of us opting blocks in to the gap support

+1 to this as well, this is a beneficial feature without axial support so I believe we should get it in and then iterate with axial support after.

stacimc committed:
Go back to using gap instead of margins

stacimc commented:
> I suppose one thing that mitigates this slightly is if users set explicit widths for the buttons, then the change in gap won't affect whether they wrap to a new line?

@andrewserong Yep, the increase in column gap will only affect wrapping for buttons that don't have a selected width, so hopefully that reduces the potential for issues there!

> For those cases where folks need a short-term work around, I suppose they could add two Buttons blocks and use margins to tweak the vertical gap?

I'm very happy you mention this -- that is definitely a viable workaround for axial spacing 😄 

> I feel somewhat strongly that we fix this in the theme, not in the block editor. A patch already fixes some of the margin issues around the navigation block. For any bugs with the Buttons block, a trac ticket and/or patch to the theme is the best path forward.

@jasmussen Thanks for this, and the example with the navigation block. This feels like a particularly gnarly issue for striking the balance between backwards compatibility and block editor tech debt, so the guidance is much appreciated! Your take makes sense to me and would allow us to do much more code clean up here.

With that in mind, the latest changes I've pushed revert back to exclusively using `gap` and setting all button margins to `0` (_without_ the `!important`). Where that leaves us:
* Twenty Twenty One still has the `first-child` issue, for which I can open a trac ticket
* Twenty Twenty has no major 'bugs', but some styling regressions: the row spacing and the gap between the Buttons container and adjacent blocks is reduced in size on the frontend. I can open a trac ticket for this as well
* Third party themes that adjust spacing with margins will need to update their styles to target `gap`
* Blocks themes such as TT1-blocks will have a slightly increased column gap by default

As others have said, I think this is a very useful even without axial support and would love to get it out there with axial controls as a future iteration.

andrewserong reviewed (commented):
Nice work @stacimc, it's looking much cleaner switching to `gap` and I love how much `red` is in the diff! I ran into an issue with the ToolsPanel unrelated to (but also revealed by) this PR while testing, and pushed a fix in https://github.com/WordPress/gutenberg/pull/35375, so you might want to give this a rebase to pick that up.

I did a bit more testing in TwentyNineteen theme, and I think we unfortunately still need one of the editor styles rule overrides, so I've left a comment about that.

I also noticed in TwentyNineteen that the button width is being overridden (for the 50% width at least) in the post editor. The style doesn't _appear_ to be coming from the theme, and I haven't been able to get to the bottom of it yet. I'm at the end of my day now, but happy to keep looking and testing tomorrow! Here's the inspect element on a button that should be 50% width in the editor (and is on the front end):

![image](https://user-images.githubusercontent.com/14988353/136151820-a32ed3e5-acd9-44f2-8755-71837e9b14ff.png)

jasmussen commented:
> I did a bit more testing in TwentyNineteen theme, and I think we unfortunately still need one of the editor styles rule overrides, so I've left a comment about that.

TwentyNineteen was one of the first themes that truly created a fully fledged editor style, and I recall it having to go to extreme lengths to create the excellent WYSIWYG effect it had at the time. 
Balancing the need to simplify editor styles for all future themes, with supporting old themes, is of course important. But notably in case of the twenty themes, I genuinely think the best path forward is to create trac tickets and fix it in the themes, especially because they can be updated separately from WordPress. 

What, particularly, is the issue in TwentyNineteen? I might be able to help there.

stacimc commented:
@andrewserong Did you mean to resolve this [comment](https://github.com/WordPress/gutenberg/pull/34546#discussion_r722920414)  about needing to add the `margin-left: 0` rule back to fix right-aligned buttons? I tested this in Twenty Nineteen and wasn't able to reproduce it. The PR should be overriding all button margins to 0 anyway, so I imagine this shouldn't be an issue unless I'm missing a problem with selector specificity.
<img width="915" alt="Screen Shot 2021-10-06 at 1 32 42 PM" src="https://user-images.githubusercontent.com/63313398/136279748-257ce1a1-a4a3-45e4-80de-48110d56c4d9.png">

> I also noticed in TwentyNineteen that the button width is being overridden (for the 50% width at least) in the post editor.

Good spot, I was able to reproduce this -- actually it looks like button widths in general are quite flaky on Twenty Nineteen. It's down to this rule in the theme's styles:

```
.editor-styles-wrapper .wp-block .wp-block {
    width: initial;
}
```

When I deselect that rule, all of the button widths immediately start working. I double-checked and it does look like this is happening on trunk as well, so at any rate it's not a regression -- looks like another selector specificity issue. Perhaps another one to push to fix in the theme?

andrewserong commented:
> Did you mean to resolve this comment about needing to add the margin-left: 0 rule back to fix right-aligned buttons?

Apologies, I didn't mean to resolve that comment! I can still reproduce the issue with the appender in TwentyTwenty and TwentyNineteen. Since you haven't been able to, I wonder if there's something different about our testing environments? I've added a video to the comment thread, but happy to do some more digging if it looks like it's just my environment, of course!

> The PR should be overriding all button margins to 0 anyway

The problem appears to be specifically with the appender `+` button in the editor, which doesn't get targeted by the `> .wp-block.wp-block-button` rule so its margin styles are still being applied from `classic.css`. Its class is `.block-list-appender wp-block`.

> When I deselect that rule, all of the button widths immediately start working. I double-checked and it does look like this is happening on trunk as well, so at any rate it's not a regression -- looks like another selector specificity issue

Thanks for looking into that! Yes, for any of the issues that are currently appearing on trunk for a particular theme, let's raise them separately so that we don't hold up this PR 🙂

stacimc committed:
Override margins for the block appender as well as the button

stacimc commented:
> The problem appears to be specifically with the appender + button in the editor

@andrewserong My bad, I see what you're saying! I'm not sure why I wasn't seeing this, but I recently destroyed/rebuilt my env and now I can repro. Mysterious! 🤷‍♀️ At any rate, I added the margin rule back in -- it looks somewhat odd because we need both rules for applying it to `.wp-block` and to `.wp-block.wp-block-button` because of specificity issues in other themes, so I added some comments. I'm still pretty happy with the amount of code clean-up here!

Thanks for all your help theme testing 😄

andrewserong commented on packages/block-library/src/buttons/style.scss:
```diff
@@ -21,16 +19,7 @@ $blocks-block__margin: 0.5em;
-			/*rtl:ignore*/
-			margin-right: 0;
-		}
+		margin: 0;
```
Unfortunately, it looks like this rule causes a styling regression on the front end of a site running TwentyTwenty theme, as its vertical margin on individual button blocks is overridden:

| Before | After |
| ---- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/136505915-01280985-ea6f-48b4-895f-e39968f052a2.png) | ![image](https://user-images.githubusercontent.com/14988353/136505971-68ad310b-c1a5-4a67-a6d0-ae599b97f18c.png) |

Is this one we should defer to the theme, rather than handling in this PR?

jasmussen replied:
In questions like these, the best way in most cases is too create a Trac ticket with a suggested fix and have it fixed in the theme, IMO.

stacimc replied:
I agree this is unfortunate but I think it's one of the fundamental tradeoffs we were hoping to make by going back to using `row-gap` instead of 'manually' setting margins to achieve row spacing 😞 Trying to solve this problem was actually one of the reasons I initially reverted back to margins in an earlier iteration.

In order to start using `row-gap` properly, we remove all the margin top/bottom styles that were previously establishing that row spacing. The tradeoff is that this reduces the gap between the entire Buttons container and adjacent blocks, because unlike margin, row-gap isn't added above/below the first/last rows.

For _blocks_ themes, I added the margin block support to the Buttons container to make this easily configurable. For classic themes, I thought about adding extra margin to the container to compensate -- but if the margin is on the container, it collapses into margin of adjacent blocks. It's another tricky one -- if we add margins to the individual buttons in addition to row-gap, we'll end up artificially increasing the spacing between rows, and the only other idea I had was adding another wrapping div (👎).

andrewserong replied:
Thanks for exploring all the options and explaining the trade-offs in detail @stacimc! That satisfies me when it comes to this issue, and I think the benefits are worth it. I'm supportive of proceeding with this PR and opening Trac tickets where needed as Joen has suggested 👍

andrewserong reviewed (commented):
This is testing really well now @stacimc! Thanks for fixing up the issue with the appender, in the process it's actually also fixed a subtle vertical alignment issue with the appender... nice!

| Before | After |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/136506288-0531f74b-abe1-40db-9d30-97ca6e68bb0d.png) | ![image](https://user-images.githubusercontent.com/14988353/136506304-be49603b-cf27-4002-ad03-e6fd1a267c57.png) |

This PR looks very close to land to me, I think it strikes a good balance of progressing the block further and adding new control for users, while trying to minimise regressions where possible. I noted another front end styling regression in TwentyTwenty — I'm not sure if this is the sort of thing we want to try to deal with in this PR, or defer to fixing in individual themes. If it weren't for encountering that issue I'd be inclined to give this the ✅ 

Personally, I'm leaning toward treating this as something for themes to address, and think that it'd be good to land this PR fairly soon if we can. I'm curious to hear what other folks think and to see if we can get a few more 👀 on this, too.

Happy to do some more testing on Monday!

andrewserong reviewed (approved):
Great work @stacimc! This is testing really nicely for me, and we've already discussed the trade-offs for all the potential regressions I encountered during testing. There's a bug still present on this branch where the blockGap value in the UI sometimes drops out when you switch between blocks, but there's already [a fix in trunk](https://github.com/WordPress/gutenberg/pull/35375) for that, and it wasn't related to this PR.

✅ Alignment/justification options are working correctly (left, center, right, space between) and in different widths
✅ Button widths are working as expected (25%, 50%, 75%, 100%)
✅ Works well for horizontal and vertical buttons variations
✅ Tested in TT1-Blocks and for regressions in classic themes TwentyTwenty, TwentyNineteen, Seedlet, Maywood

<img src="https://user-images.githubusercontent.com/14988353/136731825-dc98fd4f-21e4-47cb-a6fe-ab36a00e5981.png" width="500" />

Tested in the editor and front end in Safari, FF, Chrome, and Edge on Mac, and on the front end on Safari and Chrome iOS, with a variety of buttons and layout configurations.

LGTM!

stacimc commented:
@andrewserong Thank you very much for sticking with this one and all your help testing! 🥇 

@jasmussen I went ahead and opened a couple trac tickets for the larger theme issues pointed out in thread, although they're not yet an issue on trunk. What do you think about moving forward with this PR -- should the theme issues be fixed first? 
* Twenty Twenty One vertically offset first button/no margin on Buttons container: https://core.trac.wordpress.org/ticket/54250
* Twenty Twenty reduced vertical spacing between buttons: https://core.trac.wordpress.org/ticket/54251

jasmussen commented:
Thank you for excellent trac tickets, and for your patience with this one. The TwentyTwenty issue actually looks like a bugfix (bug good to note), and the TT1 is a result of the first-child last-child rules which really should be fixed in the theme. 

The thing about such first-child/last-child rules is that they don't seem to consider horizontal containers where those rules cause a staircase effect. As we have more and more of those, it seems important to revisit that decision at the source. In that vein, I'd think it okay to land this one as is, **especially as the staircase effect is an editor-only issue, and not present on the frontend**. I'll see if I can find time to push a suggested fix to TT1.

One thing I noticed, however, was that the new Row block is not affected by those first-child/last-child rules:
<img width="675" alt="Screenshot 2021-10-13 at 08 44 07" src="https://user-images.githubusercontent.com/1204802/137082629-8b35edc1-d707-45a6-8fc2-f511284438ac.png">

This is because it uses the new `layout` properties to define it as a flex item, where gap is all that's used to space things out, and margins are explicitly zeroed out:

<img width="412" alt="Screenshot 2021-10-13 at 08 33 12" src="https://user-images.githubusercontent.com/1204802/137081163-b6e87a9a-6147-4701-9a19-3979a87082e7.png">

Refactoring the Buttons block to use a similar layout setup might be a nice separate exploration. But it would not fix the issue in TT1, because TT1 has explicit button rules for the first-child/last-child margins, which as noted we need to remove.

However that did remind me that now that as the Buttons block is refactored to use `gap` — there's little use for per-button margins, really. Which means we could add pretty high specificity to the rule. You already have this rule to do that:

```
	// Specificity needed in some themes to override editor auto block margins for the button.
	> .wp-block.wp-block-button {
		margin: 0;
	}
```

If we were to simply change the above to the following, it would override those TT1 rules:

```
	// Specificity needed in some themes to override editor auto block margins for the button.
	> .wp-block-button.wp-block-button.wp-block-button.wp-block-button.wp-block-button {
		margin: 0;
	}
```

Resulting in this:

<img width="733" alt="after fixes" src="https://user-images.githubusercontent.com/1204802/137081658-4679c2e0-f7f4-4a85-80e1-a4419c220e44.png">

The arbitrary amount of class duplications here is a little ugly. But because the margin of child blocks should always be zero, it feels okay to add very high specificity here, almost bordering on `!important`. So if you like, and after all (I realize you've probably already been around this solution, sorry about that), we could push that small fix. It would fix the editor view for TT1, and because margins shouldn't be used here, it should not negatively affect any other themes.

---

One more curiosity: I noticed that for TT1 specifically, the `--wp--style--block-gap` variable is present in the _editor_, but not on the frontend, meaning there's a 24px gap in the editor, and a 0.5em gap on the frontend:

<img width="764" alt="Screenshot 2021-10-13 at 08 51 04" src="https://user-images.githubusercontent.com/1204802/137082326-72047037-02af-44ba-b3f3-81aa1a7c7af0.png">

<img width="398" alt="Screenshot 2021-10-13 at 08 51 01" src="https://user-images.githubusercontent.com/1204802/137082422-0f918d3c-1b38-455c-b01e-5636f9be66c3.png">

I assume this is a separate issue, so shouldn't block this PR, but then I tested TwentyNineteen, where the `--wp--style--block-gap` is missing both in the editor and the frontend. Any idea why TT1 is special in that regard?

stacimc committed:
Further increase specificity to fix TT1 editor styles

stacimc commented:
>  But because the margin of child blocks should always be zero, it feels okay to add very high specificity here, almost bordering on !important. So if you like, and after all (I realize you've probably already been around this solution, sorry about that), we could push that small fix.

Pushed! When I originally suggested an `!important` here, there was concern about the tradeoff of shrinking the row-gap on classic themes that override it with margins (basically the Twenty Twenty issue). Since that is inevitable anyway I'm happy to make this change; the more dramatic visual bug with offset first buttons in Twenty Twenty One is now cleaned up.

 > I noticed that for TT1 specifically, the --wp--style--block-gap variable is present in the editor, but not on the frontend

Ah yes, this is caused by a bug where the variable would sometimes render in classic themes -- it was fixed in #35209 and I have now rebased just to pick that up and confirm that it's working.

jasmussen commented:
Cool, I think we should try this. Gap vastly simplifies the user experience, as well as the implementation, and the behavior is both predictable and nearly always desirable when spacing items out. This takes a big step towards that, and doesn't preclude us from revisiting and tweaking in the future, as themes start to embrace this, or in case issues get surfaced.

And thank you for your work on this!

NicoHood commented:
Is this feature available in the recent 5.9 release? It seems that I am unable to use it. It might be caused by the astra theme or the astra gutenberg block plugin, and if so, I'd report it there. Any hints?

![grafik](https://user-images.githubusercontent.com/6888294/154042426-605a3f41-9a70-4835-97d1-98b2cf56c8e4.png)

andrewserong commented:
> Is this feature available in the recent 5.9 release?

It is available in 5.9, however themes need to explicitly opt-in to the `blockGap` support (or to all UI controls via the `appearanceTools` setting) in their `theme.json` file. There's more info on this in the [block-editor handbook here](https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/#what-is-blockgap-and-how-can-i-use-it). So, it sounds like the best bet here might be to contact the theme developer to see if they're planning on adding support.

NicoHood commented:
Thanks! Is there a way to enable it for me with a code snippet in the meantime?

Do you also know how to properly set the margin after a button block? For me I always have to add a paragraph afterwards, as the following text, image, etc is sticked way to close. I know this is another issue, but I could not find any information why this happens. It seems to be an upstream issue or decision. This workaround fixes it in the frontend, but not in the editor. Maybe you got any hint?

```css
.wp-block-button {
    margin-bottom: 0.5em;
}
```

andrewserong commented:
> Is there a way to enable it for me with a code snippet in the meantime?

Unfortunately, no. The way that each theme implements support for Gutenberg features can be quite unique depending on how they're set up, so beyond what's in [the handbook](https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/#what-is-blockgap-and-how-can-i-use-it), it's probably best to contact the theme developers via their support forums if you're not building your own theme from scratch.

> Do you also know how to properly set the margin after a button block?

Usually the Button block would be inside a Buttons block, so I don't think you'd need to add a bottom margin on the individual button? Again, this sounds like it's likely to do with how the theme handles spacing. In blocks-based themes that opt-in to block gap and margin you can select the optional margin control to adjust the spacing between the Buttons block and the following paragraph for a one-off change (however ideally the theme would set its desired spacing at the global level):

![image](https://user-images.githubusercontent.com/14988353/154370176-32e51052-ffbf-479c-8aa7-1f5ccc2a181c.png)

If the theme you're using doesn't support that, the theme developers will be the best ones to let you know what flexibility they provide.

If you think there's a bug in Gutenberg, though, and it doesn't look like an issue has been reported for it, feel free to create an issue!

bradhogan commented:
Doesn't seem like there's a way to add a blockGap value to the core/button block in the theme.json, is that still an open issue?  I've seen a solution like this that would be useful so a user wouldn't have to adjust the spacing every time they add buttons - assuming the default blockGap is larger.  

```
"core/buttons": {
	"spacing": {
		"blockGap": ".5em"
	}
}
```

andrewserong commented:
Thanks for reporting @bradley2083! That's correct, the `blockGap` setting isn't currently supported at the block-level within global styles / `theme.json`, but we're hoping to add it in. I've opened up an issue in https://github.com/WordPress/gutenberg/issues/39789 to track progress.
